plugins/keyboard-navigation/src/navigation.js

```diff
@@ -380,4 +380,505 @@ export class Navigation {
+      return false;
+    }
+    if (block.getParent()) {
+      block.unplug(false);
+    }
+    block.moveTo(wsNode.getWsCoordinate());
+    return true;
+  }
+
+  /**
+   * Disconnects the child block from its parent block. No-op if the two given
+   * connections are unrelated.
+   * @param {!Blockly.RenderedConnection} movingConnection The connection that
+   *     is being moved.
+   * @param {!Blockly.RenderedConnection} destConnection The connection to be
+   *     moved to.
+   * @protected
+   */
+  disconnectChild(movingConnection, destConnection) {
+    const movingBlock = movingConnection.getSourceBlock();
+    const destBlock = destConnection.getSourceBlock();
+
+    if (movingBlock.getRootBlock() === destBlock.getRootBlock()) {
+      if (movingBlock.getDescendants(false).indexOf(destBlock) > -1) {
+        this.getInferiorConnection(destConnection).disconnect();
+      } else {
+        this.getInferiorConnection(movingConnection).disconnect();
+      }
+    }
+  }
+
+  /**
+   * Tries to connect the  given connections.
+   *
+   * If the given connections are not compatible try finding compatible
+   * connections on the source blocks of the given connections.
+   *
+   * @param {?Blockly.RenderedConnection} movingConnection The connection that
+   *     is being moved.
+   * @param {?Blockly.RenderedConnection} destConnection The connection to be
+   *     moved to.
+   * @return {boolean} True if the two connections or their target connections
+   *     were connected, false otherwise.
+   * @protected
+   */
+  connect(movingConnection, destConnection) {
+    if (!movingConnection || !destConnection) {
+      return false;
+    }
+
+    const movingInferior = this.getInferiorConnection(movingConnection);
+    const destSuperior = this.getSuperiorConnection(destConnection);
+
+    const movingSuperior = this.getSuperiorConnection(movingConnection);
+    const destInferior = this.getInferiorConnection(destConnection);
+
+    if (movingInferior && destSuperior &&
+        this.moveAndConnect(movingInferior, destSuperior)) {
+      return true;
+      // Try swapping the inferior and superior connections on the blocks.
+    } else if (
+        movingSuperior && destInferior &&
+        this.moveAndConnect(movingSuperior, destInferior)) {
+      return true;
+    } else if (this.moveAndConnect(movingConnection, destConnection)) {
+      return true;
+    } else {
+      const checker = movingConnection.getConnectionChecker();
+      const reason =
+          checker.canConnectWithReason(movingConnection, destConnection, false);
+      this.warn(
+          'Connection failed with error: ' +
+          checker.getErrorMessage(reason, movingConnection, destConnection));
+      return false;
+    }
+  }
+
+  /**
+   * Finds the inferior connection on the source block if the given connection
+   * is superior.
+   * @param {?Blockly.RenderedConnection} connection The connection trying to be
```

Comment:
This parameter is marked nullable, but in the body of the function you assume it's non-null. Ideally Closure should tell you if you've done this but maybe that's part of strict checking we don't have on yet. Most of the other functions that have nullable parameters handle it correctly but some of them don't (the next function also does this and maybe more).

Reply:
Ok, went through and I think I got all of them.